experiments/MERA/modified_binary_mera_lib.py

```diff
@@ -0,0 +1,535 @@
+                           steady state density matrices 
+    verbose:               int 
+                           verbosity flag 
+    opt_u, opt_uv:         bool 
+                           if False, skip unitary or isometry optimization 
+    numpy_update:          bool
+                           if True, use numpy svd to calculate update of disentanglers
+    opt_all_layers:        bool
+                           if True, optimize all layers
+                           if False, optimize only truncating layers
+    opt_u_after:           int 
+                           start optimizing disentangler only after `opt_u_after` initial optimization steps
+    Returns: 
+    -------------------------------
+    (wC, vC, uC, rhoAB, rhoBA, run_times, Energies)
+    wC, vC, uC:             list of tf.Tensor 
+                            obtimized MERA tensors
+    rhoAB, rhoBA:           tf.Tensor 
+                            steady state density matrices at the top layer 
+    run_times:              list 
+                            run times per iteration step 
+    Energies:               list 
+                            energies at each iteration step
+    """
+    dtype = rhoAB_0.dtype
+    
+
+    hamAB = [0 for x in range(len(vC) + 1)]
+    hamBA = [0 for x in range(len(vC) + 1)]
+    rhoAB = [0 for x in range(len(vC) + 1)]
+    rhoBA = [0 for x in range(len(vC) + 1)]
+    
+    hamAB[0] = hamAB_0
+    hamBA[0] = hamBA_0
+    
+    chi1 = hamAB[0].shape[0]
+    
+    bias = tf.math.reduce_max(tf.linalg.eigvalsh(tf.reshape(hamAB[0],(chi1 * chi1, chi1 * chi1))))
+    hamAB[0] = hamAB[0] - bias * tf.reshape(tf.eye(chi1 * chi1, dtype=dtype), (chi1, chi1, chi1, chi1))
+    hamBA[0] = hamBA[0] - bias * tf.reshape(tf.eye(chi1 * chi1, dtype=dtype), (chi1, chi1, chi1, chi1))
+    
+    skip_layer = [misc_mera.skip_layer(w) for w in wC]
+    for p in range(len(wC)):
+        if skip_layer[p]:
+            hamAB[p+1], hamBA[p+1] = ascending_super_operator(hamAB[p],hamBA[p],wC[p],vC[p],uC[p],refsym)            
+
+    Energies = []
+    run_times = []
+    for k in range(numiter):
+        t1 = time.time()
+        rhoAB_0, rhoBA_0 = steady_state_density_matrices(nsteps_steady_state, rhoAB_0, rhoBA_0, wC[-1], vC[-1], uC[-1], refsym)
+        rhoAB[-1] = rhoAB_0
+        rhoBA[-1] = rhoBA_0        
+        for p in range(len(rhoAB)-2,-1,-1):
+            rhoAB[p], rhoBA[p] = descending_super_operator(rhoAB[p+1],rhoBA[p+1],wC[p],vC[p],uC[p],refsym)
+
+        if verbose > 0:
+            if np.mod(k,10) == 1:
+                Energies.append((ncon.ncon([rhoAB[0],hamAB[0]],[[1,2,3,4],[1,2,3,4]]) + 
+                                ncon.ncon([rhoBA[0],hamBA[0]],[[1,2,3,4],[1,2,3,4]]))/4 + bias/2)
+                stdout.write('\rIteration: %i of %i: E = %.8f, err = %.16f at D = %i with %i layers' %
+                             (int(k),int(numiter), float(Energies[-1]), float(Energies[-1] + 4/np.pi,), int(wC[-1].shape[2]), len(wC)))
```

Comment:
Bring your line lengths to be less than 80 characters. (We will likely add this restriction to the presubmit soon.)